packages/@lwc/template-compiler/src/shared/renderer-hooks.ts

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import { invariant, TemplateErrors } from '@lwc/errors';
+import State from '../state';
+import { BaseElement, ElementDirective } from './types';
+
+/**
+ * Config representing criteria for an element match.
+ */
+export interface CustomRendererElementConfig {
+    /**
+     * Tag name to use to match an element.
+     */
+    tagName: string;
+    /**
+     * Optional namespace to match an element.
+     */
+    namespace?: string;
+    /**
+     * Specify attributes that need to be matched.
+     * This field is optional. When not set, the element is matched based on tag name and namespace.
+     */
+    attributes?: string[];
+}
+
+/**
+ * Config to specify which elements and directives require a customizable renderer.
+ */
+export interface CustomRendererConfig {
+    elements: CustomRendererElementConfig[];
+    directives: string[];
+    rendererModule?: string;
+}
+
+/**
+ * Mapping of Directive.name to string literals used in template
+ */
+export const LWC_DIRECTIVES: Record<ElementDirective['name'], string> = {
+    Dom: 'lwc:dom',
+    Dynamic: 'lwc:dynamic',
+    InnerHTML: 'lwc:inner-html',
+    Key: 'key',
+};
+
+export function isCustomRendererHookRequired(element: BaseElement, state: State): boolean {
+    let addCustomRenderer = false;
+    if (state.config.customRendererConfig) {
+        const { attributes, directives } = element;
+        if (directives.length) {
```

Comment:
This `directives.length` check it not strictly necessary; if it's `0`, then the `directives.some()` will always return false.

Reply:
I just mostly did it for perf and to keep the following under one relevant block for readability.
````if (addCustomRenderer) {
                // Directives that require custom renderer are not allowed on custom elements```

Reply:
It's probably an unmeasurable perf benefit, especially since this runs server-side, but if you prefer the readability, then no worries. :+1: